### PR TITLE
Separate peer set and validator set

### PR DESCRIPTION
**Pull request by tailingchen**

### CODE REVIEW

File: consensus/pbft/backends/simple/peer.go [hunk 1]
```diff
@@ -17,56 +17,85 @@
...
-	return p.publicKey != ""
+func (ps *peerSet) Add(p *peer) {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
+	if p == nil {
+		return
+	}
+
+	if ps.peers[p.ID()] == nil {
```
Comment (markya0616) [inline consensus/pbft/backends/simple/peer.go]: if the peer existed, we should replace it.

File: consensus/pbft/backend.go [hunk 2]
```diff
@@ -26,8 +26,8 @@ type Backend interface {
 	// ID returns self id
 	ID() uint64
 
-	// Peers returns all connected peers
-	Peers() PeerSet
+	// GetValidators returns validators
+	GetValidators() *ValidatorSet
```
Comment (markya0616) [inline consensus/pbft/backend.go]: Prefer to use Validators()

File: consensus/pbft/validator_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	testAddress2 = "0xb37866a925bccd69cfa98d43b510f1d23d78a851"
+)
+
+func TestValidatorSet(t *testing.T) {
+	testNormalValSet(t)
+	testEmptyValSet(t)
+}
+
+func testNormalValSet(t *testing.T) {
+	valSet := NewValidatorSet()
```
Comment (bailantaotao) [inline consensus/pbft/validator_test.go]: Is a redundant code?

File: consensus/pbft/validator.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+package pbft
+
+import (
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type Validator struct {
```
Comment (alanchchen) [inline consensus/pbft/validator.go]: Do we have plan to use `interface` here?

File: consensus/pbft/validator.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+func (valSet *ValidatorSet) GetProposer() *Validator {
+	if valSet.proposer == nil {
+		valSet.CalcProposer(0)
+	}
+	return valSet.proposer
+}
+
+func (valSet *ValidatorSet) CalcProposer(seed uint64) {
```
Comment (alanchchen) [inline consensus/pbft/validator.go]: In my imagination, I'd like to have capability of different proposer election policies. E.g., Sticky proposer, round robin, weighted round  robin, etc.
We might need to abstract this in the future.

File: consensus/pbft/backends/simple/peer.go (hunk 1 above)
Comment (tailingchen) [inline consensus/pbft/backends/simple/peer.go]: fixed in `2a49b4a`

File: consensus/pbft/backend.go (hunk 2 above)
Comment (tailingchen) [inline consensus/pbft/backend.go]: fixed in `	2a49b4a`

File: consensus/pbft/validator_test.go (hunk 3 above)
Comment (tailingchen) [inline consensus/pbft/validator_test.go]: fixed in `	2a49b4a`

File: consensus/pbft/validator.go (hunk 4 above)
Comment (tailingchen) [inline consensus/pbft/validator.go]: 恩, 之後可能會有不一樣的validator set實作, 包含挑選proposer的方法, 我還沒想好要怎麼訂interface所以這邊就先定義成struct

File: consensus/pbft/validator.go (hunk 5 above)
Comment (tailingchen) [inline consensus/pbft/validator.go]: 的確, 不過目前還沒有確定proposer election policies跟validator set的實作方法是否要綁在一起, 所以之後會再另外的ticket做這件事情

File: consensus/pbft/backends/simple/backend.go [hunk 6]
```diff
@@ -74,27 +76,29 @@ func (sb *simpleBackend) ID() uint64 {
...
+	pbftByte, err := Encode(&pbftMsg)
+	if err != nil {
+		return
+	}
+
+	// send to self
+	go sb.pbftEventMux.Post(pbftMsg)
+
+	// send to other peers
+	for _, peer := range sb.peerSet.List() {
```
Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: Send to peers or validators? In my imagination, it should be validators.

Comment (tailingchen) [inline consensus/pbft/backends/simple/backend.go]: Peer set is a set of `connected` validators. Peer has `public key` and `peer id` delivered by protocol manager, and validator only has `address` extracted from block header.

Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: OK

File: consensus/pbft/backends/simple/engine.go [hunk 7]
```diff
@@ -92,66 +93,66 @@ func (sb *simpleBackend) APIs(chain consensus.ChainReader) []rpc.API {
...
+	if peer == nil {
+		return
+	}
+
+	msgEvent, err := Decode(data)
+	if err != nil {
+		return
+	}
+
+	if val := sb.valSet.GetByAddress(peer.Address()); val != nil {
```
Comment (markya0616) [inline consensus/pbft/backends/simple/engine.go]: If the peer is from peerSet (connected validators), should we need to check whether the peer address is in `valSet` again?

File: consensus/pbft/backends/simple/backend.go [hunk 8]
```diff
@@ -74,27 +76,29 @@ func (sb *simpleBackend) ID() uint64 {
...
 }
 
-func (sb *simpleBackend) Peers() pbft.PeerSet {
-	return sb.peerSet
+func (sb *simpleBackend) Validators() *pbft.ValidatorSet {
+	return sb.valSet
 }
 
 func (sb *simpleBackend) Send(data []byte) {
-	peers := sb.peerSet.Peers()
```
Comment (yutelin) [inline consensus/pbft/backends/simple/backend.go]: We should return error for Send method. Probably add this in later tickets.

File: consensus/pbft/backends/simple/engine.go [hunk 9]
```diff
@@ -92,66 +93,66 @@ func (sb *simpleBackend) APIs(chain consensus.ChainReader) []rpc.API {
...
+	sb.peerSet.Remove(peerID)
 }
 
-func (sb *simpleBackend) HandleMsg(publicKey string, data []byte) {
-	if peer := sb.peerSet.GetByPublicKey(publicKey); peer != nil {
-		go sb.pbftEventMux.Post(pbft.MessageEvent{
-			ID:      peer.ID(),
-			Payload: data,
-		})
+func (sb *simpleBackend) HandleMsg(peerID string, data []byte) {
```
Comment (yutelin) [inline consensus/pbft/backends/simple/engine.go]: Should return error. Let's add it in later tickets too.

File: consensus/pbft/backends/simple/engine.go (hunk 7 above)
Comment (tailingchen) [inline consensus/pbft/backends/simple/engine.go]: Just for double checking.
`HandleMsg` is called by protocol manager, but the protocol manager doesn't know the peer is validator or not.
If one peer support pbft protocol but not in current validator set, protocol manager still call `HandleMsg`.

File: consensus/pbft/backends/simple/backend.go (hunk 8 above)
Comment (tailingchen) [inline consensus/pbft/backends/simple/backend.go]: yes, i will update it in another ticket.

File: consensus/pbft/backends/simple/engine.go (hunk 9 above)
Comment (tailingchen) [inline consensus/pbft/backends/simple/engine.go]: yes, i will update it in another ticket.

**Pull request merged**